server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -520,35 +521,43 @@ public ClusterState addIndexTemplateV2(final ClusterState currentState, final bo
      */
     private static void validateDataStreamsStillReferenced(ClusterState state, String templateName,
                                                            ComposableIndexTemplate newTemplate) {
-
         final Set<String> dataStreams = state.metadata().dataStreams().keySet();
 
-        // Generate a metadata as if the new template were actually in the cluster state
-        final Metadata metadataWithoutTemplate = Metadata.builder(state.metadata()).put(templateName, newTemplate).build();
-        final Set<String> dataStreamsWithoutTemplates = new HashSet<>();
-        // For each data stream that we have, see whether it's covered by a different
-        // template (which is great), or whether it's now uncovered by any template
-        for (String dataStream : dataStreams) {
-            final String matchingTemplate = findV2Template(metadataWithoutTemplate, dataStream, false);
-            if (matchingTemplate == null) {
-                dataStreamsWithoutTemplates.add(dataStream);
-            } else {
-                // We found a template that still matches, great! Buuuuttt... check whether it
-                // is a data stream template, as it's only useful if it has a data stream definition
-                if (metadataWithoutTemplate.templatesV2().get(matchingTemplate).getDataStreamTemplate() == null) {
-                    dataStreamsWithoutTemplates.add(dataStream);
+        Function<Metadata, Set<String>> findUnreferencedDataStreams = meta -> {
```

Comment:
If we read the data streams inside this function we can make it static. What do you think?

Reply:
I don't think it makes too big of a difference (the `Function` part is simply for re-using the code), if you think it would be better, I can split it into a separate static function though?